cmake/llext-edk.cmake

```diff
@@ -53,17 +35,69 @@ foreach(dir ${INCLUDE_DIRS})
         set(dest ${LLEXT_EDK_INC}/${dir})
     endif()
 
+    set(${relative_out} ${dest} PARENT_SCOPE)
+    if(to_prj_bindir)
+        set(${bindir_out} TRUE PARENT_SCOPE)
+    else()
+        set(${bindir_out} FALSE PARENT_SCOPE)
+    endif()
+endfunction()
+
+string(REGEX REPLACE "[^a-zA-Z0-9]" "_" llext_edk_name_sane ${LLEXT_EDK_NAME})
+string(TOUPPER ${llext_edk_name_sane} llext_edk_name_sane)
+set(install_dir_var "${llext_edk_name_sane}_INSTALL_DIR")
+
+separate_arguments(LLEXT_CFLAGS NATIVE_COMMAND ${LLEXT_CFLAGS})
+
+set(make_relative FALSE)
+foreach(flag ${LLEXT_CFLAGS})
+    if (flag STREQUAL "-imacros")
+        set(make_relative TRUE)
+    elseif (make_relative)
+        set(make_relative FALSE)
+        cmake_path(GET flag PARENT_PATH parent)
+        cmake_path(GET flag FILENAME name)
+        relative_dir(${parent} dest bindir)
+        cmake_path(RELATIVE_PATH dest BASE_DIRECTORY ${LLEXT_EDK} OUTPUT_VARIABLE dest_rel)
+        if(bindir)
+            list(APPEND imacros_gen_make "-imacros\$(${install_dir_var})/${dest_rel}/${name}")
+            list(APPEND imacros_gen_cmake "-imacros\${${install_dir_var}}/${dest_rel}/${name}")
+        else()
+            list(APPEND imacros_make "-imacros\$(${install_dir_var})/${dest_rel}/${name}")
+            list(APPEND imacros_cmake "-imacros\${${install_dir_var}}/${dest_rel}/${name}")
+        endif()
+    else()
+        list(APPEND new_cflags ${flag})
+    endif()
+endforeach()
+set(LLEXT_CFLAGS ${new_cflags})
+
+cmake_path(CONVERT "${INTERFACE_INCLUDE_DIRECTORIES}" TO_CMAKE_PATH_LIST INCLUDE_DIRS)
+
+set(autoconf_h_edk ${LLEXT_EDK_INC}/${AUTOCONF_H})
+cmake_path(RELATIVE_PATH AUTOCONF_H BASE_DIRECTORY ${PROJECT_BINARY_DIR} OUTPUT_VARIABLE autoconf_h_rel)
+
+list(APPEND base_flags_make ${LLEXT_CFLAGS} ${imacros_make})
+list(APPEND base_flags_cmake ${LLEXT_CFLAGS} ${imacros_cmake})
+
+file(MAKE_DIRECTORY ${LLEXT_EDK_INC})
+foreach(dir ${INCLUDE_DIRS})
+    if (NOT EXISTS ${dir})
+        continue()
+    endif()
+
+    relative_dir(${dir} dest bindir)
     # Use destination parent, as the last part of the source directory is copied as well
     cmake_path(GET dest PARENT_PATH dest_p)
 
     file(MAKE_DIRECTORY ${dest_p})
     file(COPY ${dir} DESTINATION ${dest_p} FILES_MATCHING PATTERN "*.h")
 
     cmake_path(RELATIVE_PATH dest BASE_DIRECTORY ${LLEXT_EDK} OUTPUT_VARIABLE dest_rel)
-    if(to_prj_bindir)
+    if(bindir)
```

Comment:
is it correctly understood that this `bindir` is used to ensure that the include paths from the build dir comes first in the list of includes ?

If we are not in control / aligned on the include path order when creating the edk then how are we sure that the other include paths are ordered correctly / ordered the same as they were for the Zephyr build ?

And if they are ordered identical to the Zephyr build, then why do we need special treatment of the build dir ?

Reply:
Not really - it's only to have the granular flags distinguish between flags that point to generated files and not generated ones.

Reply:
and why is that distinction important ?

There is a risk that a changed include order can impact which header files are being picked up.

For a simple sample like `hello_world` for native_sim, then Zephyr uses this order:
```
/<path-to-zephyr>/include                                                                           
/<path-to-zephyr>/build/zephyr/include/generated                                                    
/<path-to-zephyr>/soc/native/inf_clock                                                              
/<path-to-zephyr>/boards/native/native_sim                                                          
/<path-to-zephyr>/scripts/native_simulator/common/src/include                                       
/<path-to-zephyr>/scripts/native_simulator/native/src/include 
```
but with this code, that now becomes two settings, like this:
```
LLEXT_INCLUDE_CFLAGS = 
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/include                                                   
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/soc/native/inf_clock                                      
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/boards/native/native_sim                                  
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/scripts/native_simulator/common/src/include               
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/scripts/native_simulator/native/src/include
```

```
LLEXT_GENERATED_INCLUDE_CFLAGS =
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/include/generated
```

Meaning order will be changed. Most likely not an issue for this small example, but a more complex application could be affected.

I would personally prefer that include order is kept, just to not having the risk of a change in order causes hard to debug issues.

Reply:
> and why is that distinction important ?

The granular CFLAGS help for use cases such as unit tests - you could mock the generated stuff, for instance. I expect most of the time people will only use `LLEXT_CFLAGS` with the whole set of flags.

> I would personally prefer that include order is kept, just to not having the risk of a change in order causes hard to debug issues.

Fair. I've updated the PR so that for the whole set of flags, the includes order isn't changed.